File: Day053_2025/Q053_GFG.py
class Solution:
    @staticmethod
    def maxLength(s: str) -> int:
        n = len(s)
        left = right = maxlength = 0
        
        # Left-to-right scan
        for ch in s:
            if ch == '(':
                left += 1
            else:
                right += 1
            if left == right:
                maxlength = max(maxlength, 2 * right)
            elif right > left:
                left = right = 0
        
        # Reset counters for right-to-left scan
        left = right = 0
        for ch in s[::-1]:
            if ch == '(':
                left += 1
            else:
                right += 1
            if left == right:
                maxlength = max(maxlength, 2 * left)
            elif left > right:
                left = right = 0
                
        return maxlength

# Example usage:
if __name__ == "__main__":
    s = "(()))("
    print("Longest valid parentheses length:", Solution.maxLength(s))

File: Day053_2025/Q053_GFG.java
class Q053_GFG {
    static int maxLength(String s) {
        int n = s.length();
        int left = 0, right = 0, maxlength = 0;
        
        // Left-to-right scan
        for (int i = 0; i < n; i++) {
            if (s.charAt(i) == '(')
                left++;
            else
                right++;
            if (left == right)
                maxlength = Math.max(maxlength, 2 * right);
            else if (right > left)
                left = right = 0;
        }
        
        // Reset counters for right-to-left scan
        left = right = 0;
        for (int i = n - 1; i >= 0; i--) {
            if (s.charAt(i) == '(')
                left++;
            else
                right++;
            if (left == right)
                maxlength = Math.max(maxlength, 2 * left);
            else if (left > right)
                left = right = 0;
        }
        
        return maxlength;
    }
    
    public static void main(String[] args) {
        // Example usage:
        String s = "(()))(";
        System.out.println("Longest valid parentheses length: " + maxLength(s));
    }
}
